### PR TITLE
Add support for device python arguments with constructors.

### PR DESCRIPTION
**Pull request by gchanan**

### CODE REVIEW

**Comment by gchanan** [comment]
This also includes some necessary sparse tensor changes that are also in https://github.com/pytorch/pytorch/pull/5225.

File: test/test_torch.py [hunk 1]
```diff
@@ -1076,8 +1076,13 @@ def _test_dtypes(self, cpu_dtypes, cuda_dtypes, is_sparse):
...
             # no ops on torch.float16 currently, cuda.float16 doesn't work on windows
             if dtype != torch.float16:
-                out = torch._C._VariableFunctions.zeros((2, 3), dtype=dtype)
-                self.assertEqual(out.dtype, dtype)
+                if dtype.is_cuda and torch.cuda.device_count() > 1:
+                    out = torch._C._VariableFunctions.zeros((2, 3), device=1, dtype=dtype)
+                    self.assertIs(dtype, out.dtype)
+                    self.assertEqual(1, out.get_device())
+                else:
+                    out = torch._C._VariableFunctions.zeros((2, 3), dtype=dtype)
```
Comment (colesbury) [inline test/test_torch.py]: torch.zeros

**Review by colesbury — APPROVED** [review]

**Comment by apaszke** [comment]
I'm curious what happens if you do `torch.ones(2, 2, dtype=torch.float, device=1)`?

**Comment by gchanan** [comment]
```
>>> torch.ones(2, 2, dtype=torch.float, device=1)

 1  1
 1  1
[torch.FloatTensor of size (2,2)]
```

This behavior is elsewhere (not related to this PR), although I'm not sure if it's the same code path:
```
variable(0).new(device=1)

[torch.FloatTensor of size (0,)]
```

**Comment by apaszke** [comment]
That's reaaaally weird. We should probably check this and raise errors saying that the arguments are inconsistent

**Comment by gchanan** [comment]
Agreed; I filed https://github.com/pytorch/pytorch/issues/5461 to track this.

**Pull request merged**